### ISSUE TITLE
Extract - emit finish only when all writes are done

### ISSUE THREAD

**Issue by ZJONSSON** [ZJONSSON/node-unzipper#17]
It is possible that Extract emits a finish even before the writer has finished.  This can create potential issues for any subsequent code that assumes all files have been fully extracted (see https://github.com/ZJONSSON/node-unzipper/issues/15)

**Comment by owo** [comment]
I have experienced this issue as well. When I perform the 'extract to directory' example from the README, 'finish' get's emitted before the extraction has completed.

**Comment by mingyuan-xia** [comment]
+1 confirmation. Simple way to reproduce: unzip a relative large zip file and `fs.stat` an uncompressed file in the `finish` callback and then `fs.stat` it again 10 seconds later. Different sizes.

**Comment by ForzaGreen** [comment]
For me it extracts correctly the file and finishes the script, without emitting any of the events `finish`, `drain`, `end`, `error`, `close`.

**Comment by modulesio** [comment]
Is there a workaround for this? It kind of make the module unusable if it's not guaranteed everything will be extracted all the time.

**Comment by modulesio** [comment]
Seems like the problem is that entries writing and parsing are not synchronized.

This emit can fire:
https://github.com/ZJONSSON/node-unzipper/blob/master/lib/parse.js#L129

And the extractor can still be writing it out, even after the stream has closed. Seems the extraction should node emit `finish`/`close` until the entries are all consumed.

**Comment by ZJONSSON** [comment]
Thanks yeah we need to get it fixed.   PRs welcome, otherwise I will take a look soon!

**Comment by modulesio** [comment]
For anyone interested, I fixed it here: https://github.com/modulesio/node-unzipper/commit/762d23a51b4ac7a37de73dde06f95e240fde1f29

That adds another `done` event when everything is _truly_ done (which is not the case on `close`, that's the bug).

I'm not sure if that's the right approach though.

**Comment by ZJONSSON** [comment]
Nice!   We could also use `.pipe` instead of `.on('entry` as `finished` would be issued automatically - I'll throw in a branch to compare

**Comment by ZJONSSON** [comment]
actually the extract is a bit weird since it inherits `Parse` - so it's not really a transform we can pipe through.  We could use `https://www.npmjs.com/package/duplexer2` to `this.pipe(reader).pipe(transform).pipe(this)` where the transform would write each file (and pipeline would only finish when files are finished.  However I don't see any immediate need to add duplexer2.   Your solution is great, put few comments - you mind doing a PR and we'll get it in.

**Comment by ZJONSSON** [comment]
The only thing missing from your solution is that `finish` (the official event of node streams) could be fired prematurely.   But as long as you listen for `close` it should be fine

**Comment by carlossless** [comment]
It seems that @modulesio solution didn't work out for me. From what I can gather, in my case the stream is closed before the last file is extracted. For reference here's the file I'm working with: 
[lokalisation.zip](https://github.com/ZJONSSON/node-unzipper/files/1915796/lokalisation.zip). The missing file is `he.json`.

I've been poking around the codebase and found that for some reason replacing `fstream.Writter` with a regular `fs.createWriteStream` seemed to have remedied that particular case. But I cannot tell how or why.

**Comment by carlossless** [comment]
I've found that [unzip-stream](https://github.com/mhr3/unzip-stream) seems to have this solved, at least that's what I can tell from testing it.

**Comment by georgyfarniev** [comment]
Also have same problem with node 10 using code below: 

```
await new Promise((resolve, reject) => {
      fs.createReadStream(this.catalogZip)
          .pipe(unzip.Extract({ path: this.catalogDir }))
          .on('close', () => resolve())
          .on('error', (e: any) => reject(e))
})
```

similar with 

```
      fs.createReadStream(this.catalogZip)
          .pipe(unzip.Extract({ path: this.catalogDir }))
          .promise()
```

I'm trying to extract xml file from zip and approximately every 5th attempt DOM XML parser says that there's no ending tag on this file. Same code works well with unzip-stream.

**Comment by ZJONSSON** [comment]
@owo @mingyuan-xia @ForzaGreen @modulesio @carlossless @georgyfarniev thank you for taking the time to report on this issue.  Will merge and publish the candidate solution tomorrow if there are no further comments.

**Comment by ZJONSSON** [comment]
published as latest `+ unzipper@0.9.0`

**Comment by dvlato** [comment]
Hi ZJONSSON, 
Is there a similar solution available when we use our own Entry Parser ?
